5.next: port over `thephpleague/container`

LordSimal opened the pull request:
Due to our code style some BC changes had to be done (mostly return types for interface methods) to this lib to get it all green, therefore it is NOT a 1:1 replacement of `thephpleague/container`

I added it to `5.next` and not `6.x` since it can be seen as an opt-in feature to use the built-in container instead of the already existing container implementation. If not suited for a minor update this can still be easily rebased onto `6.x`

It of course needs to be connected to the rest of the framework (e.g. the Testsuite).

LordSimal committed:
port over thephpleague/container

LordSimal committed:
adjust root composer.json

dereuromark commented on src/Container/Argument/ArgumentResolverInterface.php:
```diff
@@ -0,0 +1,23 @@
+interface ArgumentResolverInterface extends ContainerAwareInterface
+{
+    /**
+     * @param array $arguments
```
```suggestion
     * @param array<LiteralArgument> $arguments
```
I guess

LordSimal replied:
phpstan told me it can be `LiteralArgument|ResolvableArgument`

LordSimal committed:
adjust stan

LordSimal committed:
enable cached auto-wiring by default

LordSimal committed:
adjust for modern PHP features

markstory commented on src/Container/Container.php:
```diff
@@ -0,0 +1,268 @@
+use Cake\Container\ServiceProvider\ServiceProviderInterface;
+use Psr\Container\ContainerInterface;
+
+class Container implements DefinitionContainerInterface
```
Would we switch the base class of `Cake\Core\Container` in a future 5.x release or leave that to 6.x?

LordSimal replied:
That would be a 6.x only move since major changes have already happened here + more major changes are going to happen according to ADmad's christmas wishlist bellow.

ADmad commented:
If we are internalizing the `league/container` package, then I would suggest making some additional changes to it too:

- Make reflection a default feature instead of requiring an additional delegate container. Apart from maybe very small projects I doubt anyone would want to declare all definitions manually.
- Provide an API to add multiple definitions. Currently the only way to add multiple definitions together is to create a `DefinitionAggregate` instance and pass it as constructor argument. It would be a lot more convenient if there was for e.g. a `Container::addDefinitions(array $definitions)` method. Then app/plugins could even have config files of their own which return the required array.
- Update `Definition::addArgument()` to have and additional parameter for argument name. Currently if I have a service who's contructor has it's 1st argument as an object and 2nd argument as a scalar, there's no way to set a value for just the 2nd argument and let the 1st argument be handled by reflection.

These are few things that bugged me when using the container.

LordSimal commented:
> Make reflection a default feature instead of requiring an additional delegate container. Apart from maybe very small projects I doubt anyone would want to declare all definitions manually.

Currently I just added the ReflectionContainer by default as you can see [here](https://github.com/cakephp/cakephp/pull/18090/commits/70e8ad69aced4d24323735b9811374a40667f109)
So you would prefer to remove this kind of ReflectionContainer abstraction as a whole?

I always thought there were reasons why one want's to manually add services to a container instead of relying on Reflection.

ADmad commented:
Disabling reflection/autowiring should be an option but available by default IMO.

LordSimal committed:
add Container::addDefinitions method

LordSimal commented on src/Container/Container.php:
```diff
@@ -88,6 +88,51 @@ public function addShared(string $id, $concrete = null): DefinitionInterface
+     * @param array<int|string, array<class-string>|class-string> $definitions
+     * @return \Cake\Container\DefinitionContainerInterface
+     */
+    public function addDefinitions(array $definitions): DefinitionContainerInterface
```
Should this be added to the interface?

dereuromark replied:
Sounds like it, yes.
This might be the only open topic left so far.

LordSimal committed:
add ability to add arguments to DIC definition via argname

LordSimal commented on src/Container/Definition/Definition.php:
```diff
@@ -276,6 +284,22 @@ protected function resolveCallable(callable $concrete): mixed
+//                    $this->addArgument($tmpArgs[$param->getName()], $param->getName());
+//                }
+//            }
+//        }
```
This was my first attempt to make partial definitions work with autowiring but I get an infinite loop doing this.

dereuromark committed:
Fix autowiring issues and CI failures for Container (#19252)

* Fix autowiring issues and add make() method for Container

- Fix interface-to-implementation binding ignoring existing definitions
  (addresses league/container#275, #278). When binding an interface to a
  concrete class that already has a registered definition, the container
  now correctly uses the existing definition instead of bypassing it.

- Add Container::make() method to allow passing constructor arguments
  during autowiring (addresses league/container#277). This allows:
  `$container->make(MyService::class, ['configValue' => 'foo'])`

- Add Container::hasDefinition() method to check if an explicit definition
  exists (vs checking if autowiring can resolve it).

- Add ReflectionContainer::getNew() to bypass caching when needed.

- Fix PHPCS issues (trailing commas in multi-line function calls).

- Fix PHPStan type errors with callable and class-string annotations.

- Fix PHPUnit 12 compatibility (replace self::returnSelf() with
  willReturnSelf()).

* Fix phpcs: remove unused imports, fix line length

* Fix missing BarInterface import in ContainerTest

* Fix PHPStan errors: unused parameter in Baz, NonExistent class reference

* Apply rector fixes and code style improvements

* Fix Container constructor to use non-nullable promoted properties

* Fix Container split package: PHP version and source URL

dereuromark commented:
All open issues fully addressed and also all https://github.com/thephpleague/container/issues open issues have been addressed along with it now.
I think we are ready here to finally move away from the unmaintained package with a clear improvement already.

LordSimal commented:
I am in favor of moving away from this external dependency and just do it ourselfs exactly how we want/need it. 👍🏻 
This lays the base work but still needs to be hooked up everywhere the current container is used

dereuromark committed:
Add addDefinitions() to DefinitionContainerInterface

dereuromark commented:
@markstory @ADmad Are we OK with the merge? We can still fine tune during 5.next dev the following months.

ADmad commented on src/Container/composer.json:
```diff
@@ -0,0 +1,48 @@
+    },
+    "replace": {
+        "orno/di": "~2.0"
+    },
```
```suggestion
```

We don't need this.

dereuromark committed:
Update src/Container/composer.json

Co-authored-by: ADmad <admad.coder@gmail.com>

dereuromark commented:
We can fine tune along the way, for now, and I checked, this seems to cover all open issues and beyond already.
